url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/QuartzScheduler.java

```diff
@@ -0,0 +1,1227 @@
+  // /
+  // / Scheduler State Management Methods
+  // /
+  // /////////////////////////////////////////////////////////////////////////
+
+  /**
+   * <p>
+   * Starts the <code>QuartzScheduler</code>'s threads that fire <code>{@link org.quartz.triggers.Trigger}s</code>.
+   * </p>
+   * <p>
+   * All <code>{@link org.quartz.triggers.Trigger}s</code> that have misfired will be passed to the appropriate TriggerListener(s).
+   * </p>
+   */
+  @Override
+  public void start() throws SchedulerException {
+
+    if (shuttingDown || closed) {
+      throw new SchedulerException("The Scheduler cannot be restarted after shutdown() has been called.");
+    }
+
+    if (initialStart == null) {
+      initialStart = new Date();
+      quartzSchedulerResources.getJobStore().schedulerStarted();
+      startPlugins();
+    }
+
+    this.quartzSchedulerThread.togglePause(false);
+
+    logger.info("Scheduler started.");
+
+    notifySchedulerListenersStarted();
+  }
+
+  @Override
+  public void startDelayed(final int seconds) throws SchedulerException {
+
+    Thread t = new Thread(new Runnable() {
+
+      @Override
+      public void run() {
+
+        try {
+          Thread.sleep(seconds * 1000L);
+        } catch (InterruptedException ignore) {
+        }
+        try {
+          start();
+        } catch (SchedulerException se) {
+          logger.error("Unable to start secheduler after startup delay.", se);
+        }
+      }
+    });
+    t.start();
+  }
+
+  /**
+   * <p>
+   * Temporarily halts the <code>QuartzScheduler</code>'s firing of <code>{@link org.quartz.triggers.Trigger}s</code>.
+   * </p>
+   * <p>
+   * The scheduler is not destroyed, and can be re-started at any time.
+   * </p>
+   */
+  @Override
+  public void standby() {
+
+    this.quartzSchedulerThread.togglePause(true);
+    logger.info("Scheduler paused.");
+    notifySchedulerListenersInStandbyMode();
+  }
+
+  /**
+   * <p>
+   * Reports whether the <code>Scheduler</code> is paused.
+   * </p>
+   */
+  @Override
+  public boolean isInStandbyMode() {
+
+    return this.quartzSchedulerThread.isPaused();
+  }
+
+  public Class getJobStoreClass() {
+
+    return quartzSchedulerResources.getJobStore().getClass();
+  }
+
+  public Class getThreadPoolClass() {
+
+    return quartzSchedulerResources.getThreadPool().getClass();
+  }
+
+  public int getThreadPoolSize() {
+
+    return quartzSchedulerResources.getThreadPool().getPoolSize();
+  }
+
+  /**
+   * <p>
+   * Halts the <code>QuartzScheduler</code>'s firing of <code>{@link org.quartz.triggers.Trigger}s</code>, and cleans up all resources associated with
+   * the QuartzScheduler.
+   * </p>
+   * <p>
+   * The scheduler cannot be re-started.
+   * </p>
+   *
+   * @param waitForJobsToComplete if <code>true</code> the scheduler will not allow this method to return until all currently executing jobs have
+   *        completed.
+   */
+  @Override
+  public void shutdown(boolean waitForJobsToComplete) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 306-354 into a separate method.